[RISCV][GISel] Add ZFA FP legalization and full tests for 9 insn

lquinn2015 opened the pull request.

mshockwave commented on llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:
```diff
@@ -582,10 +582,13 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
+                               G_INTRINSIC_ROUNDEVEN, G_FMAXIMUM, G_FMINIMUM})
+        
+      .legalFor(ST.hasStdExtZfa() /*ST.hasStdExtF()*/, {s32})
+      .legalFor(ST.hasStdExtZfa() & ST.hasStdExtD()  , {s64})
```
please use &&

lquinn2015 replied:
Sure no problem fixed in repush. I am going to rebase the changes to one commit when I figure out the minimum and maximum stuff

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:
```diff
@@ -582,10 +582,13 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
   getActionDefinitionsBuilder({G_FCEIL, G_FFLOOR, G_FRINT, G_FNEARBYINT,
                                G_INTRINSIC_TRUNC, G_INTRINSIC_ROUND,
-                               G_INTRINSIC_ROUNDEVEN})
+                               G_INTRINSIC_ROUNDEVEN, G_FMAXIMUM, G_FMINIMUM})
```
Do we have tests for libcalls for G_FMAXIMUM, G_FMINIMUM without Zfa?

lquinn2015 replied:
No we do not and you also cannot legalize these via libcall either it seems that is a defect :/  

was trying with this snippet, I suppose this change needs more work. I will shift this to a draft

llc mtriple=riscv32 -mattr+f -global-isel %s  
```
define float @fmaximum(float %a, float %b) {
   %c = call float @llvm.maximum.f32(float %a, float %b)
    ret float %c
}
```

topperc replied:
Just add a separate `getActionDefinitionsBuilder({G_FMAXIMUM, G_FMINIMUM})` that doesn't contain the libcallFor.

lquinn2015 replied:
Yeah this is the best choice for now we can hand the libcall in a different patch set

topperc commented on llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll:
```diff
@@ -0,0 +1,78 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
This tests more than the legalizer. Put it in the llvm/test/CodeGen/RISCV/GlobalISel/ directory and name it half-zfa.ll like the SelectionDAG test.

lquinn2015 replied:
Moved this I temporarly disabled the Fmaximum and Fminimum tests for consistency I think we can still test the lib call cases for s32 and s64 cases but I need to figure out the libcall pattern for it

mshockwave commented on llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll:
please test legalizer with MIR tests, just like other tests in the same folder

topperc replied:
I don't think there's much value in that at this point now that we have pretty good support for scalar. End to end IR tests cover legalizer, regbank, and isel much more efficiently.

lquinn2015 replied:
Yeah recent push moved this to  IR .ll tests legalization is just a problem in the libcall case for the min/max imum cases gotta figure that out

topperc replied:
We don't need to solve the min/max libcall in this patch. We can just enabled min/max with Zfa and leave the non-Zfa case for later.

topperc commented on llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll:
```diff
@@ -0,0 +1,78 @@
+; CHECK-ZFA-f16:       # %bb.0:
+; CHECK-ZFA-f16-NEXT:    fround.h fa0, fa0, rup
+; CHECK-ZFA-f16-NEXT:    ret
+    %b = call half @llvm.ceil.f16(half %a)
```
I believe we usually use 2 space indentation for IR

lquinn2015 replied:
yes this is true

lquinn2015 committed:
[RISCV][GISel] Add ZFA 9 insn F_(min/max)imum added to GIsel

Signed-off-by: Luke Quinn <quic_lquinn@quicinc.com>

topperc commented on llvm/test/CodeGen/RISCV/GlobalISel/float-zfa.ll:
```diff
@@ -0,0 +1,161 @@
+; RUN: | FileCheck %s --check-prefixes=CHECK-ZFA-f32
+; RUN: llc -mtriple=riscv64 -mattr=+zfa -global-isel < %s \
+; RUN: | FileCheck %s --check-prefixes=CHECK-ZFA-f32
+; RUN: llc -mtriple=riscv32 -mattr=+f -global-isel < %s \
```
We should already have coverage without Zfa in float-intrinsics.ll

lquinn2015 replied:
Removed the intrinsic check here

lquinn2015 commented:
Still failing the MIR check validaton for G_MINIMUM and G_MAXIMUM i don't know why perhaps my test isn't covering something up

topperc commented on llvm/test/CodeGen/RISCV/GlobalISel/double-zfa.ll:
```diff
@@ -0,0 +1,79 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+
+; RUN: llc -mtriple=riscv32 -mattr=+zfa,d -global-isel < %s \
+; RUN: | FileCheck %s --check-prefixes=CHECK-ZFA-f64
```
You can drop --check-prefixes and let it use the default of `CHECK`

lquinn2015 replied:
Dropped additional check prefix

michaelmaitland commented on llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:
```diff
@@ -582,12 +582,19 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
   getActionDefinitionsBuilder({G_FCEIL, G_FFLOOR, G_FRINT, G_FNEARBYINT,
                                G_INTRINSIC_TRUNC, G_INTRINSIC_ROUND,
                                G_INTRINSIC_ROUNDEVEN})
+      .legalFor(ST.hasStdExtZfa() /*ST.hasStdExtF()*/, {s32})
```
Is it intentional for ST.hasStdExtF to be commented out? I am a little confused on its meaning as a comment.

michaelmaitland replied:
I'd probably drop it, or change it to `/*hasStdExtF() is implied by hasStdExtZfa*/`

lquinn2015 replied:
I'm going to drop it, comments arent checked by the compiler its probably just more confusing to include it

lquinn2015 commented:
@topperc  - I updated the legalizer check i did not include the ZFA extension in the Sweep. Rather I added the alias opcode. Let me know if that makes sense. If we add the extension it passes the extra rules as "Oks"

topperc commented on llvm/test/CodeGen/RISCV/GlobalISel/half-zfa.ll:
```diff
@@ -0,0 +1,78 @@
+  %b = call half @llvm.roundeven.f16(half %a)
+  ret half %b
+}
+define half @fmaximum(half %a, half %b) {
```
Missing blank line here

lquinn2015 replied:
I will do better with the formatting next time! but yes this is now fixed

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:
```diff
@@ -582,12 +582,19 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
   // FIXME: We can do custom inline expansion like SelectionDAG.
-  // FIXME: Legal with Zfa.
   getActionDefinitionsBuilder({G_FCEIL, G_FFLOOR, G_FRINT, G_FNEARBYINT,
                                G_INTRINSIC_TRUNC, G_INTRINSIC_ROUND,
```
There are no tests for G_INTRINSIC_TRUNC

lquinn2015 replied:
This should be handle now. 

I have 9 instrinsics  9 tests by 3 types for each possible size

topperc reviewed (approved):
LGTM
